File: frontend/src/api/admin/upstreamPriceMonitor.ts
import { apiClient } from '../client'

export type UpstreamPriceMonitorMode = 'observe' | 'auto_apply'
export type UpstreamPriceMonitorRuntimeStatus =
  | 'idle'
  | 'running'
  | 'degraded'
  | 'failed'
  | 'disabled'
export type UpstreamPriceEvidenceStatus =
  | 'trusted'
  | 'pending'
  | 'mismatch'
  | 'stale'
  | 'unobservable'
export type UpstreamPriceEvidenceSource = 'user_request' | 'active_probe' | 'price_page'
export type UpstreamPriceDimension =
  | 'fixed_per_request'
  | 'input'
  | 'output'
  | 'cache_write'
  | 'cache_read'
  | 'per_request_lte_256k'
  | 'per_request_256k_512k'
  | 'per_request_gt_512k'
export type UpstreamPriceDimensionStatus = 'observed' | 'unobserved' | 'pending' | 'failed'
export type UpstreamPriceRunStatus =
  | 'running'
  | 'completed'
  | 'partial'
  | 'failed'

export interface UpstreamPriceMonitorConfig {
  enabled: boolean
  mode: UpstreamPriceMonitorMode
  interval_minutes: number
  markup: number
  display_multiplier_decimals: number
  account_ids: number[]
  channel_ids: number[]
  domestic_models: string[]
  per_request_models: string[]
  passive_sample_max_age_minutes: number
  active_probe_enabled: boolean
  /** Pure active mode never consumes or reconciles customer requests. */
  active_only: boolean
  active_probe_max_models_per_run: number
  active_probe_max_requests_per_model: number
  active_probe_run_budget_usd: number
  active_probe_daily_budget_usd: number
  updated_at?: string
}

export interface UpstreamPriceCoverage {
  trusted: number
  total: number
}

export interface UpstreamPriceMonitorRuntime {
  status: UpstreamPriceMonitorRuntimeStatus
  last_run_at: string | null
  next_run_at: string | null
  consecutive_failures: number
  last_error: string
  today_probe_cost: number
  current_run_probe_cost?: number
  remaining_daily_probe_budget_usd?: number
  coverage: UpstreamPriceCoverage
  current_run_id?: number | null
  key_exclusive?: boolean | null
  reconciliation_status?: string
}

export interface UpstreamPriceValues {
  fixed_per_request?: number | null
  input_per_million?: number | null
  output_per_million?: number | null
  cache_write_per_million?: number | null
  cache_read_per_million?: number | null
  per_request_lte_256k?: number | null
  per_request_256k_512k?: number | null
  per_request_gt_512k?: number | null
}

export interface UpstreamPriceUsageCounters {
  requests: number
  input_tokens: number
  output_tokens: number
  cache_creation_tokens: number
  cache_read_tokens: number
  actual_cost: number
}

export interface UpstreamPriceEvidence {
  model: string
  account_id: number
  billing_mode?: 'token' | 'per_request'
  status: UpstreamPriceEvidenceStatus
  source: UpstreamPriceEvidenceSource
  reconciliation_status?: string
  observed_at: string | null
  sample_count: number
  prices?: UpstreamPriceValues | null
  remote_delta?: UpstreamPriceUsageCounters | null
  local_delta?: UpstreamPriceUsageCounters | null
  current_prices?: UpstreamPriceValues | null
  suggested_prices?: UpstreamPriceValues | null
  display_prices_current?: UpstreamPriceValues | null
  display_multiplier_current?: number | null
  display_multiplier_suggested?: number | null
  dimension_statuses?: Partial<Record<UpstreamPriceDimension, UpstreamPriceDimensionStatus>>
  last_error?: string
}

export interface UpstreamPriceRunSummary {
  changed_models?: number
  skipped_models?: number
  failed_models?: number
  applied_models?: number
  [key: string]: unknown
}

export interface UpstreamPriceMonitorRun {
  id: number
  trigger: 'scheduled' | 'manual' | 'active_probe' | string
  status: UpstreamPriceRunStatus
  mode: UpstreamPriceMonitorMode | string
  dry_run?: boolean
  started_at: string
  finished_at?: string | null
  matched_models: number
  mismatched_models: number
  probed_models?: number
  probe_cost: number
  snapshot_hash?: string
  applied_at?: string | null
  rollback_available?: boolean
  error?: string
  summary?: UpstreamPriceRunSummary | null
}

export interface UpstreamPriceEvidenceListResponse {
  items: UpstreamPriceEvidence[]
}

export interface UpstreamPriceRunListParams {
  page?: number
  page_size?: number
  status?: UpstreamPriceRunStatus | ''
}

export interface UpstreamPriceRunListResponse {
  items: UpstreamPriceMonitorRun[]
  total: number
  page?: number
  page_size?: number
}

export interface CreateUpstreamPriceRunRequest {
  dry_run: true
  model_names?: string[]
}

/** New async handlers may return an accepted envelope; older handlers return the run directly. */
export interface CreateUpstreamPriceRunAcceptedResponse {
  accepted?: boolean
  poll_after_ms?: number
  run: UpstreamPriceMonitorRun
}

export interface ApplyUpstreamPriceRunRequest {
  snapshot_hash: string
}

export type UpstreamPriceModelStatus =
  | 'managed'
  | 'discovered'
  | 'suspected_retired'
  | 'ignored'
  | 'retired'

export interface UpstreamPriceModelCatalogEntry {
  model: string
  status: UpstreamPriceModelStatus
  domestic_candidate: boolean
  seen_account_count: number
  expected_account_count: number
  missing_runs: number
  discovery_complete: boolean
  first_seen_at?: string | null
  last_seen_at?: string | null
  last_missing_at?: string | null
  updated_at: string
}

const basePath = '/admin/upstream-price-monitor'

export async function getConfig(options?: { signal?: AbortSignal }): Promise<UpstreamPriceMonitorConfig> {
  const { data } = await apiClient.get<UpstreamPriceMonitorConfig>(`${basePath}/config`, {
    signal: options?.signal,
  })
  return data
}

export async function updateConfig(
  config: UpstreamPriceMonitorConfig,
): Promise<UpstreamPriceMonitorConfig> {
  const { data } = await apiClient.put<UpstreamPriceMonitorConfig>(`${basePath}/config`, config)
  return data
}

export async function getRuntime(options?: { signal?: AbortSignal }): Promise<UpstreamPriceMonitorRuntime> {
  const { data } = await apiClient.get<UpstreamPriceMonitorRuntime>(`${basePath}/runtime`, {
    signal: options?.signal,
  })
  return data
}

export async function listEvidence(options?: { signal?: AbortSignal }): Promise<UpstreamPriceEvidenceListResponse> {
  const { data } = await apiClient.get<UpstreamPriceEvidenceListResponse>(`${basePath}/evidence`, {
    signal: options?.signal,
  })
  return data
}

export async function listRuns(
  params: UpstreamPriceRunListParams = {},
  options?: { signal?: AbortSignal },
): Promise<UpstreamPriceRunListResponse> {
  const { data } = await apiClient.get<UpstreamPriceRunListResponse>(`${basePath}/runs`, {
    params,
    signal: options?.signal,
  })
  return data
}

export async function listModels(options?: { signal?: AbortSignal }): Promise<{ items: UpstreamPriceModelCatalogEntry[] }> {
  const { data } = await apiClient.get<{ items: UpstreamPriceModelCatalogEntry[] }>(`${basePath}/models`, {
    signal: options?.signal,
  })
  return data
}

export async function updateModelStatus(
  model: string,
  status: UpstreamPriceModelStatus,
): Promise<UpstreamPriceModelCatalogEntry> {
  const { data } = await apiClient.post<UpstreamPriceModelCatalogEntry>(`${basePath}/models/status`, { model, status })
  return data
}

export async function discoverModels(): Promise<{ items: UpstreamPriceModelCatalogEntry[] }> {
  const { data } = await apiClient.post<{ items: UpstreamPriceModelCatalogEntry[] }>(`${basePath}/models/discover`)
  return data
}

export async function createRun(
  request: CreateUpstreamPriceRunRequest = { dry_run: true },
): Promise<UpstreamPriceMonitorRun> {
  const { data } = await apiClient.post<UpstreamPriceMonitorRun | CreateUpstreamPriceRunAcceptedResponse>(
    `${basePath}/runs`,
    request,
  )
  return 'run' in data ? data.run : data
}

export async function applyRun(
  id: number,
  request: ApplyUpstreamPriceRunRequest,
): Promise<UpstreamPriceMonitorRun> {
  const { data } = await apiClient.post<UpstreamPriceMonitorRun>(
    `${basePath}/runs/${encodeURIComponent(String(id))}/apply`,
    request,
  )
  return data
}

export async function rollbackRun(
  id: number,
  request: ApplyUpstreamPriceRunRequest,
): Promise<UpstreamPriceMonitorRun> {
  const { data } = await apiClient.post<UpstreamPriceMonitorRun>(
    `${basePath}/runs/${encodeURIComponent(String(id))}/rollback`,
    request,
  )
  return data
}

export const upstreamPriceMonitorAPI = {
  getConfig,
  updateConfig,
  getRuntime,
  listEvidence,
  listRuns,
  listModels,
  updateModelStatus,
  discoverModels,
  createRun,
  applyRun,
  rollbackRun,
}

export default upstreamPriceMonitorAPI

File: backend/internal/domain/upstream_price_monitor.go
package domain

import "time"

// UpstreamPriceMonitorMode controls whether a completed reconciliation is
// observation-only, waits for administrator review, or is applied
// automatically. Enabled only controls scheduling and never broadens the
// selected mode's write authority.
type UpstreamPriceMonitorMode string

const (
	UpstreamPriceMonitorModeObserve   UpstreamPriceMonitorMode = "observe"
	UpstreamPriceMonitorModeReview    UpstreamPriceMonitorMode = "review"
	UpstreamPriceMonitorModeAutoApply UpstreamPriceMonitorMode = "auto_apply"
)

type UpstreamPriceMonitorRunTrigger string

const (
	UpstreamPriceMonitorRunTriggerScheduled UpstreamPriceMonitorRunTrigger = "scheduled"
	UpstreamPriceMonitorRunTriggerManual    UpstreamPriceMonitorRunTrigger = "manual"
	UpstreamPriceMonitorRunTriggerProbe     UpstreamPriceMonitorRunTrigger = "active_probe"
)

type UpstreamPriceMonitorRunStatus string

const (
	UpstreamPriceMonitorRunStatusRunning   UpstreamPriceMonitorRunStatus = "running"
	UpstreamPriceMonitorRunStatusCompleted UpstreamPriceMonitorRunStatus = "completed"
	UpstreamPriceMonitorRunStatusPartial   UpstreamPriceMonitorRunStatus = "partial"
	UpstreamPriceMonitorRunStatusFailed    UpstreamPriceMonitorRunStatus = "failed"
)

type UpstreamPriceEvidenceStatus string

const (
	UpstreamPriceEvidenceStatusTrusted      UpstreamPriceEvidenceStatus = "trusted"
	UpstreamPriceEvidenceStatusPending      UpstreamPriceEvidenceStatus = "pending"
	UpstreamPriceEvidenceStatusMismatch     UpstreamPriceEvidenceStatus = "mismatch"
	UpstreamPriceEvidenceStatusStale        UpstreamPriceEvidenceStatus = "stale"
	UpstreamPriceEvidenceStatusUnobservable UpstreamPriceEvidenceStatus = "unobservable"
)

type UpstreamPriceEvidenceSource string

const (
	UpstreamPriceEvidenceSourceUserRequest UpstreamPriceEvidenceSource = "user_request"
	UpstreamPriceEvidenceSourceActiveProbe UpstreamPriceEvidenceSource = "active_probe"
	UpstreamPriceEvidenceSourcePricePage   UpstreamPriceEvidenceSource = "price_page"
)

type UpstreamPriceReconciliationStatus string

const (
	UpstreamPriceReconciliationBaseline     UpstreamPriceReconciliationStatus = "baseline"
	UpstreamPriceReconciliationMatched      UpstreamPriceReconciliationStatus = "matched"
	UpstreamPriceReconciliationNoActivity   UpstreamPriceReconciliationStatus = "no_activity"
	UpstreamPriceReconciliationMismatch     UpstreamPriceReconciliationStatus = "mismatch"
	UpstreamPriceReconciliationRemoteReset  UpstreamPriceReconciliationStatus = "remote_reset"
	UpstreamPriceReconciliationMixedContext UpstreamPriceReconciliationStatus = "mixed_context"
)

// DefaultX5M5XDomesticModels is the deliberately closed token-model allowlist.
// Matching is case-insensitive, while the canonical upstream spelling is
// preserved in persisted evidence and management responses.
var DefaultX5M5XDomesticModels = []string{
	"deepseek-v4-flash-0731",
	"deepseek-v4-pro-0813",
	"deepseek-v4-flash-vision-exp",
	"kimi-k2.6",
	"kimi-k2.7-code",
	"kimi-k3",
	"glm-5.1",
	"glm-5.2",
	"glm-5.3",
	"glm-5.3-flash",
	"MiniMax-M2.7",
	"MiniMax-M2.7-highspeed",
	"MiniMax-M3",
	"qwen3.7-max",
	"qwen3.8-flash",
	"qwen3.8-max",
	"mimo-v2.5",
	"mimo-v2.5-pro",
	"hy3",
}

// DefaultX5M5XPerRequestModels is the independently monitored price-page
// catalogue. These models do not need to exist on the token-billing account:
// their authoritative three-tier prices come from the public upstream page.
var DefaultX5M5XPerRequestModels = []string{
	"Auto-Model",
	"deepseek-v4-flash-0731",
	"deepseek-v4-pro-0813",
	"glm-5.1",
	"glm-5.2",
	"glm-5.3",
	"glm-5.3-flash",
	"gpt-5.6",
	"grok-4.6",
	"kimi-k2.6",
	"kimi-k2.7-code",
	"MiniMax-M2.7",
	"MiniMax-M2.7-highspeed",
	"MiniMax-M3",
}

// UpstreamPriceMonitorConfig is the singleton persisted control-plane
// configuration. AccountIDs refer to existing production accounts; the
// monitor never stores another copy of their API keys.
type UpstreamPriceMonitorConfig struct {
	Enabled                    bool                     `json:"enabled"`
	Mode                       UpstreamPriceMonitorMode `json:"mode"`
	IntervalMinutes            int                      `json:"interval_minutes"`
	Markup                     float64                  `json:"markup"`
	DisplayMultiplierDecimals  int                      `json:"display_multiplier_decimals"`
	AccountIDs                 []int64                  `json:"account_ids"`
	ChannelIDs                 []int64                  `json:"channel_ids"`
	DomesticModels             []string                 `json:"domestic_models"`
	PerRequestModels           []string                 `json:"per_request_models"`
	PassiveSampleMaxAgeMinutes int                      `json:"passive_sample_max_age_minutes"`
	ActiveProbeEnabled         bool                     `json:"active_probe_enabled"`
	ActiveOnly                 bool                     `json:"active_only"`
	ActiveProbeMaxRequests     int                      `json:"active_probe_max_requests_per_model"`
	ActiveProbeMaxModels       int                      `json:"active_probe_max_models_per_run"`
	ActiveProbeRunBudgetUSD    float64                  `json:"active_probe_run_budget_usd"`
	ActiveProbeDailyBudgetUSD  float64                  `json:"active_probe_daily_budget_usd"`
	UpdatedAt                  time.Time                `json:"updated_at"`
}

func DefaultUpstreamPriceMonitorConfig() UpstreamPriceMonitorConfig {
	return UpstreamPriceMonitorConfig{
		Enabled:                    false,
		Mode:                       UpstreamPriceMonitorModeObserve,
		IntervalMinutes:            360,
		Markup:                     1.20,
		DisplayMultiplierDecimals:  3,
		DomesticModels:             append([]string(nil), DefaultX5M5XDomesticModels...),
		PerRequestModels:           append([]string(nil), DefaultX5M5XPerRequestModels...),
		PassiveSampleMaxAgeMinutes: 1440,
		ActiveProbeEnabled:         false,
		ActiveOnly:                 true,
		ActiveProbeMaxRequests:     7,
		ActiveProbeMaxModels:       19,
		ActiveProbeRunBudgetUSD:    0.15,
		ActiveProbeDailyBudgetUSD:  0.40,
	}
}

// UpstreamPriceUsageCounters uses the x5m5x /v1/usage accounting dimensions.
// Cost is deliberately excluded from exact-match equality: the local cost was
// computed from our current price configuration, while ActualCost is the
// upstream truth that the monitor is trying to infer.
type UpstreamPriceUsageCounters struct {
	Requests            int64   `json:"requests"`
	InputTokens         int64   `json:"input_tokens"`
	OutputTokens        int64   `json:"output_tokens"`
	CacheCreationTokens int64   `json:"cache_creation_tokens"`
	CacheReadTokens     int64   `json:"cache_read_tokens"`
	ActualCost          float64 `json:"actual_cost"`
}

type UpstreamPriceRemoteUsageSnapshot struct {
	AccountID  int64                                 `json:"account_id"`
	LedgerDate string                                `json:"ledger_date"`
	CapturedAt time.Time                             `json:"captured_at"`
	Models     map[string]UpstreamPriceUsageCounters `json:"models"`
}

// UpstreamPriceBillingSnapshot is a sanitized copy of
// /v1/sub2api/billing. It contains no key or authorization material.
type UpstreamPriceBillingSnapshot struct {
	ResolvedRateMultiplier  float64   `json:"resolved_rate_multiplier"`
	EffectiveRateMultiplier float64   `json:"effective_rate_multiplier"`
	PeakRateEnabled         bool      `json:"peak_rate_enabled"`
	PeakStart               string    `json:"peak_start,omitempty"`
	PeakEnd                 string    `json:"peak_end,omitempty"`
	PeakRateMultiplier      *float64  `json:"peak_rate_multiplier,omitempty"`
	AppliedPeakMultiplier   float64   `json:"applied_peak_multiplier"`
	Timezone                string    `json:"timezone,omitempty"`
	ObservedAt              time.Time `json:"observed_at"`
}

// UpstreamPriceLocalAggregate is read from committed usage_logs between two
// monotonically increasing ID watermarks for one account and effective
// upstream model.
type UpstreamPriceLocalAggregate struct {
	ModelName            string                     `json:"model_name"`
	Counters             UpstreamPriceUsageCounters `json:"counters"`
	FirstUsageAt         *time.Time                 `json:"first_usage_at,omitempty"`
	LastUsageAt          *time.Time                 `json:"last_usage_at,omitempty"`
	DistinctServiceTiers int                        `json:"distinct_service_tiers"`
	HasSpecialContext    bool                       `json:"has_special_context"`
}

type UpstreamPriceUsageCheckpoint struct {
	AccountID            int64                      `json:"account_id"`
	ModelName            string                     `json:"model_name"`
	AccountIdentityHash  string                     `json:"account_identity_hash"`
	Remote               UpstreamPriceUsageCounters `json:"remote"`
	LedgerDate           string                     `json:"ledger_date"`
	BillingContextHash   string                     `json:"billing_context_hash"`
	LocalUsageLogID      int64                      `json:"local_usage_log_id"`
	CapturedAt           time.Time                  `json:"captured_at"`
	ActiveProbePending   bool                       `json:"active_probe_pending"`
	ActiveProbeStartedAt *time.Time                 `json:"active_probe_started_at,omitempty"`
	Revision             int64                      `json:"revision"`
}

// UpstreamPriceVector stores USD prices per one million tokens, or absolute
// per-request tier prices. Nil means unobserved, never zero.
type UpstreamPriceVector struct {
	FixedPerRequest      *float64 `json:"fixed_per_request,omitempty"`
	InputPerMillion      *float64 `json:"input_per_million,omitempty"`
	OutputPerMillion     *float64 `json:"output_per_million,omitempty"`
	CacheWritePerMillion *float64 `json:"cache_write_per_million,omitempty"`
	CacheReadPerMillion  *float64 `json:"cache_read_per_million,omitempty"`
	PerRequestLTE256K    *float64 `json:"per_request_lte_256k,omitempty"`
	PerRequest256K512K   *float64 `json:"per_request_256k_512k,omitempty"`
	PerRequestGT512K     *float64 `json:"per_request_gt_512k,omitempty"`
}

type UpstreamPriceEvidence struct {
	ID                         int64                             `json:"id"`
	RunID                      int64                             `json:"run_id"`
	AccountID                  int64                             `json:"account_id"`
	ModelName                  string                            `json:"model"`
	BillingMode                string                            `json:"billing_mode"`
	Status                     UpstreamPriceEvidenceStatus       `json:"status"`
	Source                     UpstreamPriceEvidenceSource       `json:"source"`
	ReconciliationStatus       UpstreamPriceReconciliationStatus `json:"reconciliation_status"`
	ContextKey                 string                            `json:"context_key"`
	ObservedAt                 time.Time                         `json:"observed_at"`
	SampleCount                int                               `json:"sample_count"`
	LocalDelta                 UpstreamPriceUsageCounters        `json:"local_delta"`
	RemoteDelta                UpstreamPriceUsageCounters        `json:"remote_delta"`
	Prices                     UpstreamPriceVector               `json:"prices"`
	CurrentPrices              UpstreamPriceVector               `json:"current_prices"`
	SuggestedPrices            UpstreamPriceVector               `json:"suggested_prices"`
	DisplayPricesCurrent       UpstreamPriceVector               `json:"display_prices_current"`
	DisplayMultiplierCurrent   *float64                          `json:"display_multiplier_current,omitempty"`
	DisplayMultiplierSuggested *float64                          `json:"display_multiplier_suggested,omitempty"`
	DimensionStatuses          map[string]string                 `json:"dimension_statuses,omitempty"`
	LastError                  string                            `json:"last_error,omitempty"`
	CreatedAt                  time.Time                         `json:"created_at"`
}

type UpstreamPriceObservation struct {
	Requests            int64
	InputTokens         int64
	OutputTokens        int64
	CacheCreationTokens int64
	CacheReadTokens     int64
	ActualCost          float64
}

type UpstreamPriceMonitorRun struct {
	ID                int64                          `json:"id"`
	Trigger           UpstreamPriceMonitorRunTrigger `json:"trigger"`
	Status            UpstreamPriceMonitorRunStatus  `json:"status"`
	Mode              UpstreamPriceMonitorMode       `json:"mode"`
	DryRun            bool                           `json:"dry_run"`
	StartedAt         time.Time                      `json:"started_at"`
	FinishedAt        *time.Time                     `json:"finished_at,omitempty"`
	MatchedModels     int                            `json:"matched_models"`
	MismatchedModels  int                            `json:"mismatched_models"`
	ProbedModels      int                            `json:"probed_models"`
	ProbeCost         float64                        `json:"probe_cost"`
	SnapshotHash      string                         `json:"snapshot_hash,omitempty"`
	Summary           map[string]any                 `json:"summary,omitempty"`
	Error             string                         `json:"error,omitempty"`
	AppliedAt         *time.Time                     `json:"applied_at,omitempty"`
	RollbackAvailable bool                           `json:"rollback_available"`
}

type UpstreamPriceMonitorRunPage struct {
	Items []UpstreamPriceMonitorRun `json:"items"`
	Total int64                     `json:"total"`
}

type UpstreamPriceMonitorRuntime struct {
	Status                       string     `json:"status"`
	LastRunAt                    *time.Time `json:"last_run_at,omitempty"`
	NextRunAt                    *time.Time `json:"next_run_at,omitempty"`
	ConsecutiveFailures          int        `json:"consecutive_failures"`
	LastError                    string     `json:"last_error,omitempty"`
	TodayProbeCost               float64    `json:"today_probe_cost"`
	CurrentRunProbeCost          float64    `json:"current_run_probe_cost"`
	RemainingDailyProbeBudgetUSD float64    `json:"remaining_daily_probe_budget_usd"`
	Coverage                     struct {
		Trusted int `json:"trusted"`
		Total   int `json:"total"`
	} `json:"coverage"`
}

type UpstreamPriceModelStatus string

const (
	UpstreamPriceModelStatusManaged          UpstreamPriceModelStatus = "managed"
	UpstreamPriceModelStatusDiscovered       UpstreamPriceModelStatus = "discovered"
	UpstreamPriceModelStatusSuspectedRetired UpstreamPriceModelStatus = "suspected_retired"
	UpstreamPriceModelStatusIgnored          UpstreamPriceModelStatus = "ignored"
	UpstreamPriceModelStatusRetired          UpstreamPriceModelStatus = "retired"
)

type UpstreamPriceDiscoveredModel struct {
	ModelName         string `json:"model"`
	SeenAccountCount  int    `json:"seen_account_count"`
	DomesticCandidate bool   `json:"domestic_candidate"`
}

type UpstreamPriceModelCatalogEntry struct {
	ModelName            string                   `json:"model"`
	Status               UpstreamPriceModelStatus `json:"status"`
	DomesticCandidate    bool                     `json:"domestic_candidate"`
	SeenAccountCount     int                      `json:"seen_account_count"`
	ExpectedAccountCount int                      `json:"expected_account_count"`
	MissingRuns          int                      `json:"missing_runs"`
	DiscoveryComplete    bool                     `json:"discovery_complete"`
	FirstSeenAt          *time.Time               `json:"first_seen_at,omitempty"`
	LastSeenAt           *time.Time               `json:"last_seen_at,omitempty"`
	LastMissingAt        *time.Time               `json:"last_missing_at,omitempty"`
	UpdatedAt            time.Time                `json:"updated_at"`
}
